### ISSUE TITLE
Compiler crashes

### ISSUE THREAD

**Issue by turtleflyer** [microsoft/TypeScript#43226]
**TypeScript Version:**  ^4.2.0-dev.20210105

**Search Terms:** compiler crashes

**Code**

This is a reduced repro:
```ts
const fn = <S, K extends keyof S>(x: K, y: S): (readonly [a: K, b: () => S[K]])[] => [
  [x, () => y] as const,
];
```

**Expected behavior:**

The `tsc` compiler must report the error:
```
error TS2322: Type 'readonly [K, () => S]' is not assignable to type 'readonly [a: K, b: () => S[K]]'.
  The types returned by '1()' are incompatible between these types.
    Type 'S' is not assignable to type 'S[K]'.

2   [x, () => y] as const,
    ~~~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

It actually does in versions prior 4.2.0-dev.20210105

**Actual behavior:**

The compiler crashes:
```
/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:87641
                throw e;
                ^

Error: Debug Failure. Unhandled Diagnostic: 2626
    at reportIncompatibleStack (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:48791:45)
    at checkTypeRelatedTo (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:48663:17)
    at elaborateElementwise (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:48072:38)
    at elaborateArrayLiteral (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:48283:28)
    at elaborateError (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:47970:28)
    at checkTypeRelatedToAndOptionallyElaborate (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:47941:32)
    at checkTypeAssignableToAndOptionallyElaborate (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:47936:20)
    at checkFunctionExpressionOrObjectLiteralMethodDeferred (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:59032:29)
    at checkDeferredNode (/home/turtlefly/.nvm/versions/node/v15.11.0/lib/node_modules/typescript/lib/tsc.js:64880:21)
    at Map.forEach (<anonymous>)
```

**Comment by giacomoalonzi** [comment]
same problem here with 4.2.3 :\

**Comment by armanio123** [comment]
Fixed in #42773
